druid-shell/src/platform/windows/keyboard.rs

```diff
@@ -0,0 +1,781 @@
+        Key::BrowserRefresh => VK_BROWSER_REFRESH,
+        Key::BrowserStop => VK_BROWSER_STOP,
+        Key::BrowserSearch => VK_BROWSER_SEARCH,
+        Key::BrowserFavorites => VK_BROWSER_FAVORITES,
+        Key::BrowserHome => VK_BROWSER_HOME,
+        Key::AudioVolumeMute => VK_VOLUME_MUTE,
+        Key::AudioVolumeDown => VK_VOLUME_DOWN,
+        Key::AudioVolumeUp => VK_VOLUME_UP,
+        Key::MediaTrackNext => VK_MEDIA_NEXT_TRACK,
+        Key::MediaTrackPrevious => VK_MEDIA_PREV_TRACK,
+        Key::MediaStop => VK_MEDIA_STOP,
+        Key::MediaPlayPause => VK_MEDIA_PLAY_PAUSE,
+        Key::LaunchMail => VK_LAUNCH_MAIL,
+        Key::LaunchMediaPlayer => VK_LAUNCH_MEDIA_SELECT,
+        Key::LaunchApplication1 => VK_LAUNCH_APP1,
+        Key::LaunchApplication2 => VK_LAUNCH_APP2,
+        Key::Alphanumeric => VK_OEM_ATTN,
+        Key::Convert => VK_CONVERT,
+        Key::ModeChange => VK_MODECHANGE,
+        Key::Process => VK_PROCESSKEY,
+        Key::Attn => VK_ATTN,
+        Key::CrSel => VK_CRSEL,
+        Key::ExSel => VK_EXSEL,
+        Key::EraseEof => VK_EREOF,
+        Key::Play => VK_PLAY,
+        Key::ZoomToggle => VK_ZOOM,
+        _ => return None,
+    })
+}
+
+fn code_unit_to_key(code_unit: u32) -> Key {
+    match code_unit {
+        0x8 | 0x7F => Key::Backspace,
+        0x9 => Key::Tab,
+        0xA | 0xD => Key::Enter,
+        0x1B => Key::Escape,
+        _ if code_unit >= 0x20 => {
+            if let Some(c) = std::char::from_u32(code_unit) {
+                Key::Character(c.to_string())
+            } else {
+                // UTF-16 error, very unlikely
+                Key::Unidentified
+            }
+        }
+        _ => Key::Unidentified,
+    }
+}
+
+/// Get location from virtual key code.
+///
+/// This logic is based on NativeKey::GetKeyLocation from Mozilla.
+fn vk_to_location(vk: VkCode, is_extended: bool) -> Location {
+    match vk as INT {
+        VK_LSHIFT | VK_LCONTROL | VK_LMENU | VK_LWIN => Location::Left,
+        VK_RSHIFT | VK_RCONTROL | VK_RMENU | VK_RWIN => Location::Right,
+        VK_RETURN if is_extended => Location::Numpad,
+        VK_INSERT | VK_DELETE | VK_END | VK_DOWN | VK_NEXT | VK_LEFT | VK_CLEAR | VK_RIGHT
+        | VK_HOME | VK_UP | VK_PRIOR => {
+            if is_extended {
+                Location::Standard
+            } else {
+                Location::Numpad
+            }
+        }
+        VK_NUMPAD0 | VK_NUMPAD1 | VK_NUMPAD2 | VK_NUMPAD3 | VK_NUMPAD4 | VK_NUMPAD5
+        | VK_NUMPAD6 | VK_NUMPAD7 | VK_NUMPAD8 | VK_NUMPAD9 | VK_DECIMAL | VK_DIVIDE
+        | VK_MULTIPLY | VK_SUBTRACT | VK_ADD | VK_ABNT_C2 => Location::Numpad,
+        _ => Location::Standard,
+    }
+}
+
+impl KeyboardState {
+    /// Create a new keyboard state.
+    ///
+    /// There should be one of these per window. It loads the current keyboard
+    /// layout and retains some mapping information from it.
+    pub fn new() -> KeyboardState {
```

Comment:
can this be `pub(crate)`? here and below

Reply:
The reason *this* is not pub(crate) is that I want to keep the same code in win-win, where it's publicly exported. But maybe that should be a diff.